### ISSUE TITLE
TimeDistributed doesn't mask correctly in some cases.

### ISSUE THREAD

**Issue by jmhessel** [keras-team/keras#13176]
**System information**  
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  1.1.4
- Keras version:  2.2.4-tf
- Python version:  3 

**Describe the current behavior**  

If a batch size is specified explicitly, TimeDistributed doesn't seem to properly pass on the mask. For example, in these lines, it seems like mask doesn't get used at all...

https://github.com/keras-team/keras/blob/master/keras/layers/wrappers.py#L209-L230

The mask seems to be handled properly for the usual case --- i.e., if you have batch shape as None.

**Describe the expected behavior**  

The mask should be properly passed on in TimeDistributed.

**Code to reproduce the issue**  
Provide a reproducible test case that is the bare minimum necessary to generate the problem.  

```python
import tensorflow as tf
tf.enable_eager_execution()
import numpy as np
np.random.seed(1)

DIM=2
LENGTH=2
SEGMENT=2
BATCH=4

my_gru = tf.keras.layers.GRU(DIM)

mask = tf.constant(
    (np.random.random((BATCH, SEGMENT, LENGTH)) > .5).astype(np.float32))
values = tf.constant(
    (np.random.random((BATCH, SEGMENT, LENGTH, DIM)).astype(np.float32)))

out_td_impl = tf.keras.layers.TimeDistributed(my_gru)(values, mask=mask)
print(mask)
print(out_td_impl)

def time_distributed_rnn_reshape(rnn, values, mask):
  num_segments = tf.shape(mask)[-1]
  values_shape = tf.shape(values)
  mask_shape = tf.shape(mask)
  flat_values = tf.reshape(values, (-1, values_shape[-2], values_shape[-1]))
  flat_mask = tf.reshape(mask, (-1, mask_shape[-1]))
  res_values = rnn(flat_values, mask=flat_mask)
  res_values = tf.reshape(res_values, [values_shape[0], values_shape[1], -1])
  return res_values

out_reshape_impl = time_distributed_rnn_reshape(my_gru, values, mask)
print(out_reshape_impl)
#out_td_impl should match out_reshape_impl, but it doesn't.
#the correct output without time distributing can be derived using the following:
#(and, it matches the reshape implementation)

#check correct output...
test_values, test_mask = values[0:1,0,:], mask[0:1,0]
test_out = my_gru(test_values, mask=test_mask)
print(test_out)
```

**Comment by jmhessel** [comment]
BTW --- perhaps a quick fix to this is simply to remove the rnn implementation of TimeDistributed... is there a performance point to it? It seems like the reshape implementation would work fine in both cases

**Comment by jmhessel** [comment]
@jvishnuvardhan Sorry to ping you about this, but do you think I should do a PR? This issue could potentially be quite problematic for running on TPU where you have to make batch size explicit.

**Comment by jmhessel** [comment]
sorry to ping you @fchollet ; but do you think I should do a PR here? Or is this the intended behavior of timedistributed? Happy to help, but wanted to check first.
